[mlir][emitc] Arith to EmitC conversion: constants

TinaAMD opened the pull request:
* Add a conversion from `arith.constant` to `emitc.constant`.
* Drop the translation for `arith.constant`s.

TinaAMD committed:
[mlir][emitc] Arith to EmitC conversion: constants

* Add a conversion from `arith.constant` to `emitc.constant`.
* Drop the translation for `arith.constant`s.

marbre reviewed (changes requested):
I just send #84151, which adds what I already had for Arith to EmitC. The changes are structured a little bit different (e.g. with the patch it is allowed to register the patterns within other passes) and further adds a Bazel build configuration. I would therefore highly appreciate if you could apply your changes op top, in particular the conversion pattern for `arith.constant`, after #84151 has landed.

marbre commented on mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp:
```diff
@@ -0,0 +1,104 @@
+
+namespace {
+
+static bool isConvertibleToEmitC(Type type) {
```
This is a rather broad check which could also be reused in other conversions. Furthermore, it mainly models a restriction which is introduced by the emitter but not the dialect itself, where both can change. Thus I think this check at least needs to be move to a more accessible location to be reused by other conversions and ideally verifiers or even the emitter. And validation pass could be another alternative, at least until the EmitC dialect and the emitter match more closely.

TinaAMD replied:
Yes, I agreed the check should be used more widely. I propose we do the following:
1) We restrict the types in the EmitC dialect (tablegen) by a function like this, which aligns with what the emitter currently supports. Later on, we could consider to restrict the types further if it makes sense. I can create a separate PR for this.
2) We drop the check in the conversion for now, and after 1) lands, we enable it again in the conversion such that the conversion rejects invalid types rather than generating invalid EmitC.
What do you think?

marbre replied:
This is exactly what I have proposed offline to @simon-camp. So yeah, sounds good to me :+1:

marbre commented on mlir/lib/Conversion/ArithToEmitC/ArithToEmitC.cpp:
```diff
@@ -0,0 +1,104 @@
+    RewritePatternSet patterns(&getContext());
+    populateArithToEmitCConversionPatterns(patterns);
+
+    if (failed(applyPartialConversion(getOperation(), target,
+                                      std::move(patterns)))) {
+      signalPassFailure();
+    }
```
Please see https://llvm.org/docs/CodingStandards.html#don-t-use-braces-on-simple-single-statement-bodies-of-if-else-loop-statements.

marbre reviewed (changes requested):
Please including dropping https://github.com/llvm/llvm-project/blob/03588a27261f7ebea15af49268d2ec901fe1979e/utils/bazel/llvm-project-overlay/mlir/BUILD.bazel#L1771 in your PR.

Regarding the other changes, I need to free up some time to take a closer look, especially at the test.

TinaAMD committed:
Drop arith dependency in bazel build

TinaAMD commented:
> Please including dropping
> 
> https://github.com/llvm/llvm-project/blob/03588a27261f7ebea15af49268d2ec901fe1979e/utils/bazel/llvm-project-overlay/mlir/BUILD.bazel#L1771
> in your PR.
> 
> Regarding the other changes, I need to free up some time to take a closer look, especially at the test.

Thanks, I dropped it!

simon-camp commented on mlir/test/Target/Cpp/stdops.mlir:
```diff
@@ -1,37 +1,6 @@
 
-func.func @std_constant() {
-  %c0 = arith.constant 0 : i32
-  %c1 = arith.constant 2 : index
-  %c2 = arith.constant 2.0 : f32
-  %c3 = arith.constant dense<0> : tensor<i32>
-  %c4 = arith.constant dense<[0, 1]> : tensor<2xindex>
-  %c5 = arith.constant dense<[[0.0, 1.0], [2.0, 3.0]]> : tensor<2x2xf32>
```
Can you convert these into `emitc.constants`and add them [here](https://github.com/llvm/llvm-project/blob/9e0f5909d0af3911b19bb1f97fb400c3ce431f63/mlir/test/Target/Cpp/const.mlir#L11)

TinaAMD replied:
See [cc76a04](https://github.com/llvm/llvm-project/pull/83798/commits/cc76a04941c981bec2f7c929881e97186b24e02a).

simon-camp reviewed (changes requested):
Can you please readd the dropped tests in `const.mlir` file. Besides that this looks good.

TinaAMD committed:
Move tests for float/tensor/index constants

Add tests for float/tensor/index types to the `emitc.constant` tests (they were previously tested on `arith.constant`s).

marbre reviewed (approved):
LGTM, thanks!

marbre commented:
@TinaAMD let me know if I should hit the "Squash and merge" button for you.

marbre commented:
> > @TinaAMD let me know if I should hit the "Squash and merge" button for you.
> 
> Yes, that would be great.

I've merged your PR but unfortunately it was associated with an anonymous GH email (`<...@users.noreply.github.com>`). Normally, there should have been run an action which points our that this isn't desired and to change the settings within your GH profile, see for example https://github.com/llvm/llvm-project/pull/83386#issuecomment-1970548438. Anyway, please make sure to change the settings accordingly for future PRs.

TinaAMD commented:
> > > @TinaAMD let me know if I should hit the "Squash and merge" button for you.
> > 
> > 
> > Yes, that would be great.
> 
> I've merged your PR but unfortunately it was associated with an anonymous GH email (`<...@users.noreply.github.com>`). Normally, there should have been run an action which points our that this isn't desired and to change the settings within your GH profile, see for example [#83386 (comment)](https://github.com/llvm/llvm-project/pull/83386#issuecomment-1970548438). Anyway, please make sure to change the settings accordingly for future PRs.

I changed it, thanks for the hint.
